third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp

```diff
@@ -112,6 +112,95 @@ warpsPerTileWMMA(tt::DotOp dotOp, const ArrayRef<int64_t> shape, int numWarps) {
+    return cast<RankedTensorType>(v.getType()).getElementType();
+  });
+
+  // Use simple costmodel to define optimal set of the dot operands.
+  // Most expensive - accuracy loss conversions:
+  //   - any larger type -> any smaller type;
+  //   - float -> int;
+  //   - int -> float (not supported for now);
+  //   - signed int -> unsigned int;
+  //   - unsigned int -> signed int with same or less size.
+  // They are never performed, better to use FMA.
+  // Supported conversion for now costs `1`, no conversion costs `0`.
+  // The model could be improved in the future. For example taken into account
+  // chain dot could be detected and result conversion score is decreased.
+  int maxConvertCost =
+      std::numeric_limits<int32_t>::max() / applicableTypes.front().size();
+  auto calcConvertCost = [&](Type fromTy, Type toTy) -> int32_t {
+    if (fromTy == toTy)
+      return 0;
+
+    // Skip conversion between int and float. Int16/int32 cases are lowered to
+    // FMA.
+    if (fromTy.isIntOrIndex() != toTy.isIntOrIndex())
+      return maxConvertCost;
+
+    if (fromTy.isIntOrIndex() && toTy.isIntOrIndex() &&
+        fromTy.isUnsignedInteger() != toTy.isUnsignedInteger())
+      return fromTy.isUnsignedInteger() && fromTy.getIntOrFloatBitWidth() <
+                                               toTy.getIntOrFloatBitWidth()
+                 ? 1
+                 : maxConvertCost;
+
+    return fromTy.getIntOrFloatBitWidth() <= toTy.getIntOrFloatBitWidth()
+               ? 1
+               : maxConvertCost;
+  };
+  auto minCost = maxConvertCost;
+  auto optTypes = OperandTypesVector();
+  for (auto types : applicableTypes) {
+    assert(types.size() == initElemTypes.size());
+    int accumulatedConvertCost = 0;
+    for (int i = 0; i < initElemTypes.size(); ++i) {
+      accumulatedConvertCost += calcConvertCost(initElemTypes[i], types[i]);
+    }
+    if (accumulatedConvertCost < minCost) {
+      minCost = accumulatedConvertCost;
+      optTypes = types;
+    }
+  }
+  return optTypes;
+}
+
+OperandTypesVector getOperandTypesForWmmaOp(mlir::PatternRewriter &rewriter,
+                                            tt::DotOp dot) {
+  Type f16 = rewriter.getF16Type();
+  Type f32 = rewriter.getF32Type();
+  Type bf16 = rewriter.getBF16Type();
+  Type si4 = rewriter.getIntegerType(4, true);
+  Type ui4 = rewriter.getIntegerType(4, false);
+  Type si8 = rewriter.getIntegerType(8, true);
+  Type ui8 = rewriter.getIntegerType(8, false);
+  Type i32 = rewriter.getIntegerType(32);
+  SmallVector<OperandTypesVector> applicableTypes = {
+      // clang-format off
+      {f16, f16, f32, f32},
+      {f16, f16, f16, f16},
+      {bf16, bf16, f32, f32},
+      {bf16, bf16, bf16, bf16},
+      {si8, si8, i32, i32},
+      {si8, ui8, i32, i32},
+      {ui8, si8, i32, i32},
+      {ui8, ui8, i32, i32},
+      {si4, si4, i32, i32},
+      {si4, ui4, i32, i32},
+      {ui4, si4, i32, i32},
+      {ui4, ui4, i32, i32},
```

Comment:
These types are not supported by Triton. I'm not sure if they are supported in torch. So how do we test the correctness of these input types?